File: Inflearn/2_탐색과 시뮬레이션/스도쿠 검사.py
import sys
board = [list(map(int, sys.stdin.readline().split())) for _ in range(9)]
dx = [0, 1, 2, 0, 1, 2, 0, 1, 2]
dy = [0, 0, 0, 1, 1, 1, 2, 2, 2]


def check(board):
    for i in range(9):
        if (len(set(board[i])) != len(board[i])):
            return "NO"
        temp = []
        for j in range(9):
            temp.append(board[j][i])
        if (len(set(temp)) != len(temp)):
            return "NO"
    for i in range(0, 7, 3):
        for j in range(0, 7, 3):
            temp = []
            for k in range(9):
                temp.append(board[i + dx[k]][j + dy[k]])
            if (len(set(temp)) != len(temp)):
                return "NO"
    return "YES"


print(check(board))

#ref
sys.stdin = open("input.txt", "r")


def check(a):
    for i in range(9):
        ch1 = [0]*10
        ch2 = [0]*10
        for j in range(9):
            ch1[a[i][j]] = 1
            ch2[a[j][i]] = 1
        if sum(ch1) != 9 or sum(ch2) != 9:
            return False
    for i in range(3):
        for j in range(3):
            ch3 = [0]*10
            for k in range(3):
                for s in range(3):
                    ch3[a[i*3+k][j*3+s]] = 1
            if sum(ch3) != 9:
                return False
    return True


a = [list(map(int, input().split())) for _ in range(9)]
if check(a):
    print("YES")
else:
    print("NO")

File: JavaScript/Inflearn/py_2_탐색과 시뮬레이션/스도쿠 검사.js
// 각 행, 열, 3x3 보드 내 1~9 중복없이 채워져있는지 확인
// 정확하게 풀었으면 YES 아니면 NO 출력
const solution = (board) => {
  const YES = 'YES';
  const NO = 'NO';
  const len = board.length;

  let numSet;
  for (let i = 0; i < len; i++) {
    numSet = new Set(board[i]);
    if (numSet.size !== len) return NO;
  }

  let check;
  for (let i = 0; i < len; i++) {
    check = Array.from({ length: len + 1 }, () => 0);
    for (let j = 0; j < len; j++) {
      if (check[board[i][j]] === 1) return NO;
      check[board[i][j]]++;
    }
  }

  for (let i = 0; i < len; i += 3) {
    for (let j = 0; j < len; j += 3) {
      check = Array.from({ length: len + 1 }, () => 0);
      for (let k = i; k < i + 3; k++) {
        for (let s = j; s < j + 3; s++) {
          if (check[board[k][s]] === 1) return NO;
          check[board[k][s]]++;
        }
      }
    }
  }

  return YES;
};

console.log(
  solution([
    [1, 4, 3, 6, 2, 8, 5, 7, 9],
    [5, 7, 2, 1, 3, 9, 4, 6, 8],
    [9, 8, 6, 7, 5, 4, 2, 3, 1],
    [3, 9, 1, 5, 4, 2, 7, 8, 6],
    [4, 6, 8, 9, 1, 7, 3, 5, 2],
    [7, 2, 5, 8, 6, 3, 9, 1, 4],
    [2, 3, 7, 4, 8, 1, 6, 9, 5],
    [6, 1, 9, 2, 7, 5, 8, 4, 3],
    [8, 5, 4, 3, 9, 6, 1, 2, 7],
  ])
); // YES
